File: typescript/src/handler/executor.ts
import { logger } from "../logger.js";
import {
  type RelayRequest,
  type RelayResponse,
  RelayResponseType,
  type Response,
  parseRequest,
  serializeResponse,
} from "../schemas.js";
import type { HandlerFn } from "./decorator.js";

class TimeoutError extends Error {}

export async function executeHandler(
  fn: HandlerFn,
  request: RelayRequest
): Promise<RelayResponse> {
  try {
    const body =
      typeof request.body === "object" && request.body !== null && !Array.isArray(request.body)
        ? (request.body as Record<string, unknown>)
        : {};
    const handlerRequest = parseRequest(body);

    const result = await withTimeout(
      Promise.resolve().then(() => fn(handlerRequest)),
      request.timeout
    );

    const response = normalize(result);
    return {
      id: request.id,
      type: RelayResponseType.RESPONSE,
      statusCode: 200,
      headers: { "Content-Type": "application/json" },
      body: serializeResponse(response),
    };
  } catch (e) {
    if (e instanceof TimeoutError) {
      logger.error(
        `handler timed out after ${request.timeout}s (request ${request.id})`
      );
      return errorResponse(
        request.id,
        `handler timed out after ${request.timeout}s`
      );
    }
    const message = e instanceof Error ? e.message : String(e);
    logger.error(`handler raised for request ${request.id}: ${message}`);
    return errorResponse(request.id, message || typeName(e));
  }
}

async function withTimeout<T>(promise: Promise<T>, seconds: number): Promise<T> {
  let timer: NodeJS.Timeout | undefined;
  const timeout = new Promise<never>((_, reject) => {
    timer = setTimeout(() => reject(new TimeoutError()), seconds * 1000);
  });
  try {
    return await Promise.race([promise, timeout]);
  } finally {
    clearTimeout(timer);
  }
}

function normalize(result: unknown): Response {
  if (isResponse(result)) {
    return result;
  }
  if (typeof result === "string") {
    return { output: result };
  }
  if (result === null || result === undefined) {
    throw new Error("handler returned null — return a string or a Response");
  }
  throw new Error(
    `handler returned ${typeName(result)} — return a string or a Response`
  );
}

function isResponse(result: unknown): result is Response {
  return (
    typeof result === "object" &&
    result !== null &&
    typeof (result as Response).output === "string"
  );
}

function typeName(value: unknown): string {
  if (typeof value === "object" && value !== null) {
    return value.constructor?.name ?? "object";
  }
  return typeof value;
}

function errorResponse(requestId: string, message: string): RelayResponse {
  return {
    id: requestId,
    type: RelayResponseType.ERROR,
    statusCode: 500,
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ error: message }),
  };
}

File: rust/src/handler/executor.rs
use std::collections::HashMap;
use std::fmt;
use std::time::Duration;

use serde_json::{json, Value};

use crate::schemas::{RelayRequest, RelayResponse, RelayResponseType, Request};

use super::decorator::HandlerFn;

#[derive(Debug)]
pub struct HandlerError(String);

impl HandlerError {
    pub fn new(message: impl Into<String>) -> Self {
        HandlerError(message.into())
    }
}

impl fmt::Display for HandlerError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}

impl std::error::Error for HandlerError {}

impl From<String> for HandlerError {
    fn from(message: String) -> Self {
        HandlerError(message)
    }
}

impl From<&str> for HandlerError {
    fn from(message: &str) -> Self {
        HandlerError(message.to_string())
    }
}

pub async fn execute_handler(f: &HandlerFn, request: &RelayRequest) -> RelayResponse {
    let handler_request = match &request.body {
        Some(body @ Value::Object(_)) => match serde_json::from_value::<Request>(body.clone()) {
            Ok(handler_request) => handler_request,
            Err(e) => {
                log::error!("handler raised for request {}: {e}", request.id);
                return error_response(&request.id, &e.to_string());
            }
        },
        _ => Request::default(),
    };

    match tokio::time::timeout(Duration::from_secs(request.timeout), f(handler_request)).await {
        Ok(Ok(response)) => match serde_json::to_string(&response) {
            Ok(body) => RelayResponse {
                id: request.id.clone(),
                r#type: RelayResponseType::Response,
                status_code: 200,
                headers: HashMap::from([(
                    "Content-Type".to_string(),
                    "application/json".to_string(),
                )]),
                body,
            },
            Err(e) => {
                log::error!("handler raised for request {}: {e}", request.id);
                error_response(&request.id, &e.to_string())
            }
        },
        Ok(Err(e)) => {
            log::error!("handler raised for request {}: {e}", request.id);
            error_response(&request.id, &e.to_string())
        }
        Err(_) => {
            log::error!(
                "handler timed out after {}s (request {})",
                request.timeout,
                request.id
            );
            error_response(
                &request.id,
                &format!("handler timed out after {}s", request.timeout),
            )
        }
    }
}

fn error_response(request_id: &str, message: &str) -> RelayResponse {
    RelayResponse {
        id: request_id.to_string(),
        r#type: RelayResponseType::Error,
        status_code: 500,
        headers: HashMap::from([("Content-Type".to_string(), "application/json".to_string())]),
        body: json!({ "error": message }).to_string(),
    }
}
